File: program1.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

class Product {
    String name;
    double price;
    int quantity;

    public Product(String name, double price) {
        this.name = name;
        this.price = price;
        this.quantity = 0;
    }
}

class Catalogue {
    Map<String, Product> products;

    public Catalogue() {
        this.products = new HashMap<>();
    }

    public void addProduct(Product product) {
        this.products.put(product.name, product);
    }

    public double getProductPrice(String name) {
        return this.products.get(name).price;
    }

    public int getProductQuantity(String name) {
        return this.products.get(name).quantity;
    }
}

public class ShoppingCart {
    public static void main(String[] args) {
        Catalogue catalogue = new Catalogue();
        catalogue.addProduct(new Product("Product A", 20));
        catalogue.addProduct(new Product("Product B", 40));
        catalogue.addProduct(new Product("Product C", 50));

        Map<String, Integer> cartQuantity = new HashMap<>();
        double cartTotal = 0;

        Scanner scanner = new Scanner(System.in);
        for (String productName : catalogue.products.keySet()) {
            System.out.print("Enter quantity for " + productName + ": ");
            int quantity = scanner.nextInt();
            cartQuantity.put(productName, quantity);
            cartTotal += quantity * catalogue.getProductPrice(productName);
        }

        String discountRule = applyDiscount(cartTotal, cartQuantity);

        double subtotal = cartTotal;
        double discountAmount = 0;
        if (discountRule != null) {
            double discount = (subtotal * Double.parseDouble(discountRule.split("_")[1])) / 100;
            subtotal -= discount;
            discountAmount = discount;
        }

        int shippingFee = (cartQuantity.values().stream().mapToInt(Integer::intValue).sum() / 10) * 5;
        int giftWrapFee = cartQuantity.values().stream().mapToInt(Integer::intValue).sum();

        double total = subtotal + shippingFee + giftWrapFee;

        System.out.println("\n=== Order Details ===");
        for (Map.Entry<String, Integer> entry : cartQuantity.entrySet()) {
            String productName = entry.getKey();
            int quantity = entry.getValue();
            System.out.println(productName + ": Quantity - " + quantity + ", Total - " + quantity * catalogue.getProductPrice(productName));
        }
        System.out.println("\nSubtotal: " + subtotal);
        System.out.println("Discount applied: " + (discountRule != null ? discountRule : "No discount applied"));
        System.out.println("Shipping fee: " + shippingFee);
        System.out.println("Gift wrap fee: " + giftWrapFee);
        System.out.println("\nTotal: " + total);
    }

    private static String applyDiscount(double cartTotal, Map<String, Integer> cartQuantity) {
        if (cartTotal > 200) {
            return "flat_10_discount";
        } else if (cartQuantity.values().stream().anyMatch(value -> value > 10)) {
            return "bulk_5_discount";
        } else if (cartQuantity.values().stream().mapToInt(Integer::intValue).sum() > 20) {
            return "bulk_10_discount";
        } else if (cartQuantity.values().stream().mapToInt(Integer::intValue).sum() > 30 &&
                cartQuantity.values().stream().anyMatch(value -> value > 15)) {
            return "tiered_50_discount";
        } else {
            return null;
        }
    }
}

File: program1.py
class ProductInfo:
    def __init__(self, name, price):
        self.name = name
        self.price = price
        self.quantity = 0

class ProductCatalogue:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        self.products[product.name] = product

    def get_product_price(self, name):
        return self.products[name].price

    def get_product_quantity(self, name):
        return self.products[name].quantity

def apply_discount_rule(cart_total, cart_quantity):
    discount_rules = {
        "flat_10_discount": (cart_total > 200, 10),
        "bulk_5_discount": (any(cart_quantity[name] > 10 for name in cart_quantity), 5),
        "bulk_10_discount": (sum(cart_quantity.values()) > 20, 10),
        "tiered_50_discount": (sum(cart_quantity.values()) > 30 and any(cart_quantity[name] > 15 for name in cart_quantity), 50)
    }

    applicable_discounts = {rule: discount for rule, (is_applicable, discount) in discount_rules.items() if is_applicable}
    if not applicable_discounts:
        return None, 0

    max_discount_rule = max(applicable_discounts, key=applicable_discounts.get)
    return max_discount_rule, applicable_discounts[max_discount_rule]

def main_shopping_cart():
    catalogue = ProductCatalogue()
    catalogue.add_product(ProductInfo("Product A", 20))
    catalogue.add_product(ProductInfo("Product B", 40))
    catalogue.add_product(ProductInfo("Product C", 50))

    cart_quantity = {}
    cart_total = 0
    for product_name in catalogue.products.keys():
        quantity = int(input(f"Enter quantity for {product_name}: "))
        cart_quantity[product_name] = quantity
        cart_total += quantity * catalogue.get_product_price(product_name)

    discount_rule, discount_amount = apply_discount_rule(cart_total, cart_quantity)

    subtotal = cart_total
    if discount_rule:
        discount = (subtotal * discount_amount) / 100
        subtotal -= discount

    shipping_fee = (sum(cart_quantity.values()) // 10) * 5
    gift_wrap_fee = sum(cart_quantity.values())

    total = subtotal + shipping_fee + gift_wrap_fee

    print("\n=== Order Details ===")
    for product_name, quantity in cart_quantity.items():
        print(f"{product_name}: Quantity - {quantity}, Total - {quantity * catalogue.get_product_price(product_name)}")
    print(f"\nSubtotal: {subtotal}")
    print(f"Discount applied: {discount_rule} - {discount_amount}%") if discount_rule else print("No discount applied")
    print(f"Shipping fee: {shipping_fee}")
    print(f"Gift wrap fee: {gift_wrap_fee}")
    print(f"\nTotal: {total}")

if __name__ == "__main__":
    main_shopping_cart()
